sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java

```diff
@@ -565,21 +611,25 @@ public static Object convertAvroFormat(
       case BOOLEAN:
         return convertAvroPrimitiveTypes(beamFieldTypeName, avroValue);
       case DATETIME:
-        // Expecting value in microseconds.
-        switch (options.getTruncateTimestamps()) {
-          case TRUNCATE:
-            return truncateToMillis(avroValue);
-          case REJECT:
-            return safeToMillis(avroValue);
-          default:
-            throw new IllegalArgumentException(
-                String.format(
-                    "Unknown timestamp truncation option: %s", options.getTruncateTimestamps()));
+        if (avroFieldSchema != null
+            && avroFieldSchema.getLogicalType() == org.apache.avro.LogicalTypes.timestampMillis()) {
+          return new Instant((long) avroValue);
```

Comment:
I think it would be better if we could map timestamp micros and timestamp millis to different Beam types, and then we could differentiate them without referencing the source avro schema. Ideally the Beam schema would leave no ambiguity.

@alexvanboxel has proposed adding a nanosecond instant logical type in https://github.com/apache/beam/pull/10486. It seems we need logical types for millis and micros instants in that same vein.

To me this seems like a good argument for deprecating the DATETIME primitive in favor of millis/micros/nanos logical types. It would provide much more clarity in situations like this.

cc @reuvenlax